### PR TITLE
Fix for Multi-day weekly recurring event missing first instance

### LINKED ISSUE #300: Multi-day weekly recurring event misses first instance
I've run into a problem with weekly recurring events that happen on multiple days, where the first instance falls on a Sunday:

```js
const icalJs = require('ical.js');
const vevent = new icalJs.Event(new icalJs.Component(icalJs.parse(`
BEGIN:VCALENDAR
VERSION:2.0
CALSCALE:GREGORIAN
BEGIN:VEVENT
DTSTART;VALUE=DATE:20170423
RRULE:FREQ=WEEKLY;BYDAY=SU,WE
SEQUENCE:0
SUMMARY:Recurring weekly on Sundays and Wednesdays
END:VEVENT
END:VCALENDAR
`)).getFirstSubcomponent('vevent'));
const iterator = vevent.iterator();

let next;
let num = 0;
while ((next = iterator.next()) && num++ < 5) {
    console.log(vevent.getOccurrenceDetails(next).startDate.toICALString());
}
```

Actual output:

```
20170426
20170430
20170503
20170507
20170510
```

Expected output (verified with Google Calendar and Apple's iCal):

```
20170423
20170426
20170430
20170503
20170507
```

It works if I add `;WKST=SU` to the `RRULE` property, but that shouldn't be necessary. Also, the instance on April 23rd will be included if I change `BYDAY=SU,WE` to `BYDAY=SU`. So it's the presence of Wednesday in the set that throws ical.js off.

Perhaps related to: https://github.com/mozilla-comm/ical.js/issues/294

### PR DESCRIPTION
**Pull request by arunpandianp**
This fixes #300 

`ICAL.Time` by default has `Sunday` as the week start and `ICAL.Recur` treats `Monday` as the default start of week. 

The change is to get dayOfWeek relative to the `rrule.wkst` before comparing. As Sunday > Monday inside a week if wkst is Monday. 

While this fixes #300, there are still places where I see similar logic being used. I'll try to look in to them.

### CODE REVIEW

**Comment by arunpandianp** [comment]
@mschroeder @kewisch can you take a look?

File: test/recur_iterator_test.js [hunk 1]
```diff
@@ -839,6 +839,29 @@ suite('recur_iterator', function() {
           '2015-01-01T08:00:00'
         ]
       });
+
+      testRRULE('FREQ=WEEKLY;COUNT=3;BYDAY=MO,SU,TU', {
```
Comment (mschroeder) [inline test/recur_iterator_test.js]: You should add comments above the testcases to describe the scenarios explicitly tested.

File: lib/ical/recur.js [hunk 2]
```diff
@@ -379,26 +379,43 @@
    * into a numeric value of that day.
    *
    * @param {String} string     The iCalendar day name
+   * @param {ICAL.Time.weekDay=} aWeekStart
```
Comment (mschroeder) [inline lib/ical/recur.js:382]: The documented argument name does not match the method signature. `aWeekStart` should be used everywhere to be consistent.

File: lib/ical/recur.js [hunk 3]
```diff
@@ -379,26 +379,43 @@
...
+      dow += 7;
+    }
+    return dow;
   };
 
   /**
    * Convert a numeric day value into its ical representation (SU, MO, etc..)
    *
    * @param {Number} num        Numeric value of given day
+   * @param {ICAL.Time.weekDay=} aWeekStart
```
Comment (mschroeder) [inline lib/ical/recur.js:397]: The documented argument name does not match the method signature. `aWeekStart` should be used everywhere to be consistent.

**Comment by mschroeder** [comment]
@arunpandianp, I'll review your pull request over the weekend.

File: lib/ical/time.js [hunk 4]
```diff
@@ -225,10 +225,13 @@
 
     /**
      * Calculate the day of week.
+     * @param {ICAL.Time.weekDay=} aWeekStart
+     *        The week start weekday, defaults to SUNDAY
      * @return {ICAL.Time.weekDay}
      */
-    dayOfWeek: function icaltime_dayOfWeek() {
-      var dowCacheKey = (this.year << 9) + (this.month << 5) + this.day;
+    dayOfWeek: function icaltime_dayOfWeek(aFirstDayOfWeek) {
```
Comment (mschroeder) [inline lib/ical/time.js]: This function should get some more unit tests now with the optional argument.

File: lib/ical/time.js [hunk 5]
```diff
@@ -225,10 +225,13 @@
 
     /**
      * Calculate the day of week.
+     * @param {ICAL.Time.weekDay=} aWeekStart
```
Comment (mschroeder) [inline lib/ical/time.js:228]: The documented argument name does not match the method signature. `aWeekStart` should be used to be consistent in the file.

File: lib/ical/time.js [hunk 6]
```diff
@@ -246,8 +249,13 @@
         h += 5;
       }
 
-      // Normalize to 1 = sunday
+      // Normalize to 1 = wkst
       h = ((h + 6) % 7) + 1;
```
Comment (mschroeder) [inline lib/ical/time.js]: The normalization step with variable first day of week can be written in a single line as `h = ((h + 7 - firstDow) % 7) + 1;`.

File: lib/ical/recur.js [hunk 7]
```diff
@@ -379,26 +379,43 @@
...
    * @return {String}           The ICAL day value, e.g SU,MO,...
    */
-  ICAL.Recur.numericDayToIcalDay = function toIcalDay(num) {
+  ICAL.Recur.numericDayToIcalDay = function toIcalDay(num, aFirstDayOfWeek) {
     //XXX: this is here so we can deal with possibly invalid number values.
     //     Also, this allows consistent mapping between day numbers and day
     //     names for external users.
-    return REVERSE_DOW_MAP[num];
+    var firstDow = aFirstDayOfWeek || ICAL.Time.SUNDAY;
+    var dow = num;
```
Comment (mschroeder) [inline lib/ical/recur.js]: This can be written in a single statement: `var dow = ((num - firstDow + 7) % 7) + 1;`

File: lib/ical/recur.js [hunk 8]
```diff
@@ -379,26 +379,43 @@
...
    * @return {Number}           Numeric value of given day
    */
-  ICAL.Recur.icalDayToNumericDay = function toNumericDay(string) {
+  ICAL.Recur.icalDayToNumericDay = function toNumericDay(string, aFirstDayOfWeek) {
     //XXX: this is here so we can deal
     //     with possibly invalid string values.
-
-    return DOW_MAP[string];
+    var firstDow = aFirstDayOfWeek || ICAL.Time.SUNDAY;
+    var dow = DOW_MAP[string];
```
Comment (mschroeder) [inline lib/ical/recur.js]: This can be written in a single statement: `return ((DOW_MAP[string] - firstDow + 7) % 7) + 1;`

**Comment by mschroeder** [comment]
@arunpandianp, I cannot finish the review today, so I will get back to you during the week, but I already started to look into your changes.

**Comment by arunpandianp** [comment]
@mschroeder thanks :)

File: lib/ical/recur_iterator.js [hunk 9]
```diff
@@ -843,11 +843,11 @@ ICAL.RecurIterator = (function() {
         this.last.month = next.month;
     },
 
-    ruleDayOfWeek: function ruleDayOfWeek(dow) {
+    ruleDayOfWeek: function ruleDayOfWeek(dow, aFirstDayOfWeek) {
```
Comment (mschroeder) [inline lib/ical/recur_iterator.js]: `aWeekStart` should be used everywhere to be consistent. Maybe you want to add documentation to this function.

File: lib/ical/recur.js [hunk 10]
```diff
@@ -379,26 +379,43 @@
...
   /**
    * Convert a numeric day value into its ical representation (SU, MO, etc..)
    *
    * @param {Number} num        Numeric value of given day
+   * @param {ICAL.Time.weekDay=} aWeekStart
+   *        The week start weekday, defaults to SUNDAY
    * @return {String}           The ICAL day value, e.g SU,MO,...
    */
-  ICAL.Recur.numericDayToIcalDay = function toIcalDay(num) {
+  ICAL.Recur.numericDayToIcalDay = function toIcalDay(num, aFirstDayOfWeek) {
```
Comment (mschroeder) [inline lib/ical/recur.js]: This function should get some more unit tests now with the optional argument.

File: lib/ical/recur.js [hunk 11]
```diff
@@ -379,26 +379,43 @@
    * into a numeric value of that day.
    *
    * @param {String} string     The iCalendar day name
+   * @param {ICAL.Time.weekDay=} aWeekStart
+   *        The week start weekday, defaults to SUNDAY
    * @return {Number}           Numeric value of given day
    */
-  ICAL.Recur.icalDayToNumericDay = function toNumericDay(string) {
+  ICAL.Recur.icalDayToNumericDay = function toNumericDay(string, aFirstDayOfWeek) {
```
Comment (mschroeder) [inline lib/ical/recur.js]: This function should get some more unit tests now with the optional argument.

**Review by mschroeder — CHANGES_REQUESTED** [review]
Sorry for taking so long to review your pull request. I have some comments you should consider. Also, feel free to contact me or start a discussion.

**Comment by arunpandianp** [comment]
@mschroeder addressed your comments.

**Comment by kewisch** [comment]
I've manually rebased this on master and fixed it in 7b8d59d6cded33d08da617ca42696755fadd6587. Thanks for the patch!

**Pull request closed without merging**